packages/contracts-bedrock/src/universal/OptimismMintableERC20FactoryInterop.sol

```diff
@@ -0,0 +1,100 @@
+    error OnionAlreadyPeeled();
+
+    /// @notice Thrown when the tokens length to verify and store mismatch.
+    error TokensLengthMismatch();
+
+    /// @notice Thrown when the computed hash onion with the given tokens and starting inner layer doesn't match the
+    /// hashOnion.
+    error InvalidHashOnion();
+
+    /// @notice Emitted when a deployment is stored after verifying the tokens against the hashOnion.
+    /// @param localToken  Address of the local token.
+    /// @param remoteToken Address of the remote token.
+    event DeploymentStored(address indexed localToken, address indexed remoteToken);
+
+    /// @notice Storage slot where hashOnion is stored at.
+    /// keccak256(abi.encode(uint256(keccak256("hashOnion")) - 1)) & ~bytes32(uint256(0xff));
+    bytes32 internal constant HASH_ONION_SLOT = 0xae706b81046b35e941cb9737b644ec489aa4690fffd2c3f501ccbed8fd5b0400;
+
+    /// @notice Initial layer on the hash onion.
+    /// keccak256(abi.encode(0));
+    bytes32 internal constant INITIAL_ONION_LAYER = 0x290decd9548b62a8d60345a988386fc84ba6bc95484008f6362f93160ef3e563;
+
+    /// @notice Semantic version.
+    /// @custom:semver +interop
+    function version() public view override returns (string memory) {
+        return string.concat(super.version(), "+interop");
+    }
+
+    /// @notice hashOnion getter.
+    /// @return hashOnion_ The hashOnion value.
+    function hashOnion() public view returns (bytes32 hashOnion_) {
+        assembly {
+            hashOnion_ := sload(HASH_ONION_SLOT)
+        }
+    }
+
+    /// @notice Verifies the hashOnion and stores the provided addresses.
+    ///         Stores the deployments of the provided tokens if the computed hashOnion is valid. It can be considered
+    ///         complete when the hashOnion matches the initial layer.
+    /// @param _localTokens  Array of local token addresses.
+    /// @param _remoteTokens Array of remote token addresses.
+    /// @param _startingInnerLayer The starting inner layer of the hashOnion.
+    function verifyAndStore(
+        address[] calldata _localTokens,
+        address[] calldata _remoteTokens,
+        bytes32 _startingInnerLayer
+    )
+        external
+    {
+        bytes32 _hashOnion = hashOnion();
+        if (_hashOnion == 0) revert HashOnionNotSet();
+        if (_hashOnion == INITIAL_ONION_LAYER) revert OnionAlreadyPeeled();
+        if (_localTokens.length != _remoteTokens.length) revert TokensLengthMismatch();
+
+        // Unpeel the hash onion and store the deployments
+        bytes32 innerLayer = _startingInnerLayer;
+        for (uint256 i; i < _localTokens.length; i++) {
+            innerLayer = keccak256(abi.encodePacked(innerLayer, abi.encodePacked(_localTokens[i], _remoteTokens[i])));
+            deployments[_localTokens[i]] = _remoteTokens[i];
+            emit DeploymentStored(_localTokens[i], _remoteTokens[i]);
```

Comment:
Should we emit also the `innerLayer` ? Idk if it has any value

Reply:
Good point, no strong opinion about it but sounds like a good data to add to the event.

Reply:
I have my doubts of this. 
I think the goal of that event is informing the `deployments` mapping was updated with new tokens. The `innerLayer` is just the calculated proof for it was done right, but I don't see to much benefits to emit this information (also, it adds more gas to a quite gas-consuming operation)

Reply:
yeah, I don't think it adds any value.
I do see more value on emitting the updated hashOnion value

Reply:
Will add this event on both, the verify and store function and the setter!